File: android/sdk/src/main/java/app/organicmaps/sdk/routing/roadshield/RoadShield.java
package app.organicmaps.sdk.routing.roadshield;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public final class RoadShield
{
  @NonNull
  public final RoadShieldType type;
  @NonNull
  public final String text;
  @Nullable
  public final String additionalText;

  // Used by JNI.
  @Keep
  private RoadShield(@NonNull RoadShieldType type, @NonNull String text, @Nullable String additionalText)
  {
    this.type = type;
    this.text = text;
    this.additionalText = additionalText;
  }

  @Override
  @NonNull
  public String toString()
  {
    return "RoadShield{type=" + type + ", text='" + text + "', additionalText='" + additionalText + "'}";
  }
}

File: android/sdk/src/main/cpp/app/organicmaps/sdk/routing/roadshield/RoadShield.cpp
#include "RoadShield.hpp"

#include "app/organicmaps/sdk/routing/roadshield/RoadShieldType.hpp"

jobject ToJavaRoadShield(JNIEnv * env, ftypes::RoadShield const & roadShield)
{
  static jclass const klass = jni::GetGlobalClassRef(env, "app/organicmaps/sdk/routing/roadshield/RoadShield");
  // clang-format off
  static jmethodID const ctorRouteInfoID = jni::GetConstructorID(env, klass,
    "("
    "Lapp/organicmaps/sdk/routing/roadshield/RoadShieldType;"  // type
    "Ljava/lang/String;"                                       // text
    "Ljava/lang/String;"                                       // additionalText
    ")V"
  );
  // clang-format on

  // clang-format off
  jobject const result = env->NewObject(klass, ctorRouteInfoID,
    ToJavaRoadShieldType(env, roadShield.m_type),
    jni::ToJavaString(env, roadShield.m_name),
    jni::ToJavaString(env, roadShield.m_additionalText)
  );
  // clang-format on
  ASSERT(result, (jni::DescribeException()));
  return result;
}
